### ISSUE TITLE
Add MCP server and instructions for LLMs

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#8676]
### What is the problem you're trying to solve?

I'd like Stylelint to provide:

- an MCP server so that coding assistants like GitHub Copilot can efficiently use Stylelint
- instructions for LLMs so they generate code out-of-the-box that adheres to the conventions in our standard config

Without an MCP server, coding assistants can be trained to be _confidently wrong_ about Stylelint's rules and configurations.

For example, when asked what is wrong with:

```css
.className {
  color: red; 

  background: red; 
}
```

It flags the camelcase class name but not the empty line between the declarations:

> The class selector .className does not match the expected pattern (should be lowercase and hyphen-separated, e.g., .class-name). No other errors or warnings are expected for this file.

When pushed about the empty line, it doubled down:

> With the standard Stylelint config (stylelint-config-standard), an empty line before the second declaration inside a rule (like between `color: red`; and `background: red;`) is not required and does not trigger a warning or error.
> Stylelint only requires empty lines between rules, not between declarations within a rule.
> So, the only warning you will get is for the class selector naming (selector-class-pattern).

---

Without being provided some instructions, coding assistants will often generate code that doesn't follow the conventions of our standard config.

For example, when prompted to "Add a media query for when the viewport is bigger than 600px" it used the prefix notation:

```css
@media (min-width: 600px) {
  .className {
    /* Add responsive styles here */
  }
}
```

### What solution would you like to see?

For us to provide a:

- Stylelint MCP Server
- list of instructions that correspond to our standard config

---

I was about to start writing a MCP Server myself, but saw that @ryo-manba [has already beaten me to it](https://www.npmjs.com/package/stylelint-mcp) 😆 .

If @ryo-manba is okay with it, I suggest we move the repo and npm package to the Stylelint GitHub and npm organisations respectively, and then document it in our [user guide](https://stylelint.io/user-guide/customize#using-stylelint).

When I prompted Copilot to recheck the example above, but to use the MCP Server this time, it was able to detect both the kebab-case class name and the empty line between the declarations. Copilot was useful in providing explanations for the warnings, too:

> selector-class-pattern: The class selector .className does not match the expected pattern (should be lowercase and hyphen-separated, e.g., .class-name).

> declaration-empty-line-before: There is an unexpected empty line before the background: red; declaration. The standard config does not allow empty lines between declarations within a rule. Remove the empty line to fix this warning.

(ESLint [recently added](https://eslint.org/blog/2025/05/eslint-v9.26.0-released/) a [MCP Server](https://eslint.org/docs/latest/use/mcp))

---

To complement the MCP server, we can provide a series of instructions for each of the rules in our standard config, e.g.:

- Use media queries in a "range context" rather than use the "min / max prefixes", e.g. `@media (width >= 10em) {}`
- Use kebab-case for all names, including custom properties, class names, layer names etc, e.g. `class-name { --custom-property: foo; }`

When I provided these instructions to Copilot and reprompted it to add a media query for when the viewport is bigger than 600px, it correctly generated:

```css
@media (width > 600px) {
  .class-name {
    /* Add responsive styles here */
  }
}
```

(The Angular team [provide a similar set of instructions](https://angular.dev/ai/develop-with-ai#custom-prompts-and-system-instructions) for their users.)

**Comment by Mouvedia** [comment]
**concerning the server**

If it has been tested to be compatible with these LLMs:

- OpenAI
- Claude
- Mistral
- Grok 4
- Kimi K2

Then I don't mind putting it under \@stylelint.

**concerning the instructions**

We can continue this discussion on https://github.com/stylelint/stylelint-config-standard
Personally I am not interested in it because I don't use it.

**Comment by jeddy3** [comment]
Thank you for chiming in @Mouvedia.

> If it has been tested to be compatible with these LLMs

Let's decide if housing an official server within our organisation and documenting it in our user guide, regardless of its current compatibility status, has value for our users. We can, if necessary, expand the server's compatibility after it's moved to the organisation.

I think there is value, as it'll have more contributors and users' eyes on it, but let's see what everyone else thinks.

> We can continue this discussion on https://github.com/stylelint/stylelint-config-standard

Over time, we've more tightly coupled our standard config into the Stylelint documentation, e.g. in [our getting started guide](https://stylelint.io/user-guide/get-started#linting-css) and our [list of rules](https://stylelint.io/user-guide/rules), to improve the onboarding experience for new users.

Let's continue the discussion here, as it may be most useful for our users to include the instructions in the user guide. And let's allow others the opportunity to chime in on these related proposals before we consider splitting them.

> Personally I am not interested in it because I don't use it.

You're welcome to skip discussions about aspects of Stylelint and its ecosystem that don't interest you.

**Comment by ryo-manba** [comment]
@jeddy3 Thanks for opening this issue!

I originally built the MCP server just to see whether there was real demand. 
After releasing it, the project did not get much attention, probably because users had no clear way to find it, so I left it in a minimal state.

I’m happy to:

* move the repo and npm package under the **stylelint** org,
* improve the documentation, and
* maintain it as an official part of the Stylelint ecosystem.

Publishing a clear set of LLM instructions, also sounds like a good idea.
That should help coding assistants follow our conventions out of the box.

In short, I fully agree with your proposal 😄

**Comment by romainmenke** [comment]
I am a bit conflicted about this but not opposed.

I think that any tool based on LLM's is a net negative.
They are harmful over such a wide range of aspects and topics (environment, privacy, security, personal agency, brain activity, ...)

Any effort to integrate with such tools feels weird and wrong to me.

But on the other hand many people are now forced to use such tools regardless of what LLM's are and their effects.

Given the enormous potential for security breaches through the MCP protocol and that people are forced to use LLM's I think it is better if we have an official integration. I think this minimizes the chance for bad outcomes 😅

**Comment by jeddy3** [comment]
I share your sentiments about LLMs, @romainmenke.

It feels like the genie is out of the bottle, though, and, like you said, we can try to minimise bad outcomes by having an official integration. As well as demonstrating the value of Stylelint in an LLM world.

Additionally, we can caveat the integration with warnings in our documentation about the potential negative effects of LLMs, e.g., that it's more energy-efficient and environmentally friendly to run Stylelint yourself than to have an LLM run it on your behalf and interpret the results.

Likewise, providing instructions means that LLMs are more likely to get code right the first time, avoiding reprompts and rounds of costly rewriting to adhere to conventions.

We'll be stepping into controversial waters, and we'll need a considered and light touch in our documentation to avoid alienating any users. However, getting ahead of this seems prudent.

**Comment by ybiquitous** [comment]
This attempt sounds so interesting to me. I agree with porting the MCP package into our organization. 👍

**Comment by ryo-manba** [comment]
If there are no objections, I’d like to transfer the repo.
I tried, but I don’t have permission.
Could someone with access help update the settings?

**Comment by jeddy3** [comment]
If you add me to the GitHub repo and the npm package, I should be able to transfer them. 

I'll wait a bit longer before actually doing the transfer, though, so that the discussion remains open a bit longer.

**Comment by ryo-manba** [comment]
@jeddy3 
Sounds good! I've sent you invites to both.

**Comment by thibaudcolas** [comment]
It’s a bit different from the other scenarios discussed here but thought it was worth mentioning – I’d love to see ways to provide better instructions to LLMs when it comes to maintaining a shared configuration, and upgrading Stylelint. The specific things I’ve tried to figure out:

- What’s the latest Stylelint, and what versions are there between "current project" and "latest"?
- What upgrades are there migration guides for?

The versions questions can be answered either with a `npm view stylelint versions --json`, or with the CHANGELOG. But the CHANGELOG is a hefty file if you’re fetching the contents naïvely. 154k tokens with the published HTML, 70k tokens with the r.jina.ai [published HTML -> Markdown translation](https://r.jina.ai/https://stylelint.io/CHANGELOG/) 72k tokens from the Markdown source in GitHub.

For the migrations guide, I believe the only option is to look at the [directory index in GitHub](https://github.com/stylelint/stylelint/tree/main/docs/migration-guide).

---

If it helps, here are the two skills I’ve been working on where I needed this: [stylelint-config-torchbox .agents/skills](https://github.com/torchbox/stylelint-config-torchbox/tree/main/.agents/skills). The "shared stylelint config" isn’t too generic but should be simple to adapt for maintenance of other shared configs. The "upgrading stylelint" skill is more generic.

**Comment by jeddy3** [comment]
@thibaudcolas Thank you for sharing your scenarios. 

It's great that you're looking into ways to improve the experience of maintaining shared configurations.

We're leaning towards staying focused on creating interfaces, codebases and documentation that work best for humans. Fortunately, these often work well for LLMs, too.

For example:

- `npx stylelint --help` is just as useful to an LLM trying to figure out how to use Stylelint as it is to the human we designed it for
- our [docs](https://github.com/stylelint/stylelint/tree/main/docs) being in markdown is great for LLMs, but first and foremost, we use markdown because it makes it easier for us to write docs for people
- [releases](https://github.com/stylelint/stylelint/releases) are useful for LLMs, but we have them so humans can keep up to date and [watch them](https://docs.github.com/en/subscriptions-and-notifications/how-tos/managing-subscriptions-for-activity-on-github/viewing-your-subscriptions) 

As our GitHub repo is the source of truth for releases and (versioned) files, including our docs, and LLMs are proficient with CLI tools, one option for your use case could be to use `gh`.

For example:

```shell
gh release list --repo stylelint/stylelint --limit 100 --json tagName --jq '.[] | .tagName' | \
  sed -n '1,/17.0.0/p' | \
  xargs -I{} sh -c 'echo "## Release {}"; gh release view {} --repo stylelint/stylelint'
```

Will give you the releases since `17.0.0`. We link to the relevant migration guide markdown file in the release notes. Again, we do this to benefit people, but it helps LLMs, too:

```
## Release 17.0.0
17.0.0
github-actions[bot] released this about 1 month ago
It contains 14 breaking changes, which we've detailed in the [migrating to 17.0.0 guide]
(https://github.com/stylelint/stylelint/blob/main/docs/migration-guide/to-17.md)
```

Hopefully, that's a token-efficient way to get the information you need in your skills. 

It's also great that your upgrading skill is available at the [skills.sh repository](https://skills.sh/?q=stylelint||) so that others can use it. I anticipate that some member of the community will also publish one for writing Stylelint plugins soon, one that'd consume our [writing plugins docs](https://github.com/stylelint/stylelint/blob/main/docs/developer-guide/plugins.md).

As the broader community and ecosystem are a better fit for the fast-changing LLM world, I'll close the issue so that MCP and instructions are community endeavours. This might change over time, but let's close the issue for now.

**Comment by thibaudcolas** [comment]
I’d never seen `gh release` before, thank you! I think that will definitely help. And re designing for humans, that makes a ton of sense to me.
